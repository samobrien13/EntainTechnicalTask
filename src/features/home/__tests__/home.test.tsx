import {fireEvent, render} from '@testing-library/react-native';

import * as ReactQuery from 'react-query';
import {Home} from '..';
import NoRaces from '../__mocks__/no-upcoming-races.json';
import NextToGo from '../__mocks__/next-to-go.json';
import React from 'react';

jest.useFakeTimers().setSystemTime(new Date('2023-03-01T03:38:00Z'));

describe(Home.name, () => {
  it('should display empty state if no upcoming races', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(
      jest.fn().mockReturnValue({
        data: {...NoRaces},
        isFetching: false,
      }),
    );
    const {getByText} = render(<Home />);
    expect(getByText('No upcoming races')).toBeTruthy();
  });

  it('should display 5 races', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(
      jest.fn().mockReturnValue({
        data: {...NextToGo},
        isFetching: false,
      }),
    );
    const {getAllByTestId} = render(<Home />);

    expect(getAllByTestId('race')).toHaveLength(5);
  });

  it('should update state on icon press', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(
      jest.fn().mockReturnValue({
        data: {...NextToGo},
        isFetching: false,
      }),
    );
    const setState = jest.fn();
    jest
      .spyOn(React, 'useState')
      .mockImplementation(() => [['horses', 'dogs'], setState]);
    const {getByTestId} = render(<Home />);
    const horsesIcon = getByTestId('horses-icon');
    fireEvent.press(horsesIcon);
    expect(setState).toHaveBeenCalledWith(['dogs']);
    setState.mockClear();

    const trotsIcon = getByTestId('trots-icon');
    fireEvent.press(trotsIcon);
    expect(setState).toHaveBeenCalledWith(['horses', 'dogs', 'trots']);
  });
});
