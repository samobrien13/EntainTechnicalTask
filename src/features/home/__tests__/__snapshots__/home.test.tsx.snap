// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should display all categories correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#ff7800",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 24,
            "fontWeight": "600",
          }
        }
      >
        Next to go
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "marginRight": 2,
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="horses-icon"
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/horses.png",
            }
          }
          style={
            {
              "height": 32,
              "width": 32,
            }
          }
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "marginHorizontal": 2,
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="dogs-icon"
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/dogs.png",
            }
          }
          style={
            {
              "height": 32,
              "width": 32,
            }
          }
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "marginLeft": 2,
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="trots-icon"
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/trots.png",
            }
          }
          style={
            {
              "height": 32,
              "width": 32,
            }
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={
      <Text>
        No upcoming races
      </Text>
    }
    contentContainerStyle={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "fontSize": 18,
        "fontWeight": "400",
        "padding": 16,
      }
    }
    data={
      [
        {
          "advertised_start": "2023-03-01T03:38:00Z",
          "category_id": "161d9be2-e909-4326-8c2c-35ed71fb460b",
          "meeting_id": "fbcf5151-a241-4670-aec8-91dade110813",
          "meeting_name": "Yonkers Raceway",
          "race_form": {
            "additional_data": "{"revealed_race_info":{"track_name":"Yonkers Raceway","state":"USA","country":"USA","number":11,"race_name":"","time":"1970-01-01T00:00:00Z"}}",
            "distance": 1609,
            "distance_type": {
              "name": "Metres",
              "short_name": "m",
            },
            "track_condition": {
              "name": "Good",
              "short_name": "good",
            },
            "weather": {
              "name": "OCAST",
            },
          },
          "race_id": "0a28d73d-00e0-425a-b51d-7ccb7423d0aa",
          "race_name": "Race 11 - 1609M",
          "race_number": 11,
          "venue_country": "USA",
          "venue_id": "db5500e1-cdaa-4748-8b06-d389b13f988a",
          "venue_name": "Yonkers Raceway",
          "venue_state": "USA",
          "video": {
            "channel": {
              "channel_id": "2c184b7b-69fd-4ca8-b342-11021a010d69",
            },
          },
        },
        {
          "advertised_start": "2023-03-01T03:38:00Z",
          "category_id": "4a2788f8-e825-4d36-9894-efd4baf1cfae",
          "meeting_id": "d9d52ee8-d3d4-4de8-a4c1-62592dc4c6b9",
          "meeting_name": "Doomben",
          "race_form": {
            "additional_data": "{"revealed_race_info":{"track_name":"Doomben","state":"QLD","country":"AUS","number":1,"race_name":"","time":"1970-01-01T00:00:00Z"}}",
            "distance": 1200,
            "distance_type": {
              "name": "Metres",
              "short_name": "m",
            },
            "race_comment": "MAID TO FIT (10) was a big improver second up when she went under by just a nose at this track. She is lightly raced with upside and will be reaching her peak now so gets her chance to break through. PARIS AVENUE (12) started a hot favourite on debut back in July and while unplaced she did work home for a 0.8L fourth over 1000m at Sunshine Coast. Likely to have developed with maturity and has been readied for her return with two solid trials. OCEAN ENDEAVOUR (11) is a first-starter by Astern from the O'Dea/Hoysted yard. Has been trialling well so can make an impact. JUMP'N FLY (9) was green but chased to place on debut. Gets a longer trip and is open to improvement.",
            "track_condition": {
              "name": "Good4",
              "short_name": "good4",
            },
            "weather": {
              "name": "Fine",
            },
          },
          "race_id": "7ea5c48c-6f83-4ec0-8106-0935272b1c3a",
          "race_name": "Journey To The Jewel Mdn Plate",
          "race_number": 1,
          "venue_country": "AUS",
          "venue_id": "fd49470f-c5ab-4c85-8db8-b93ba4c62cf8",
          "venue_name": "Doomben",
          "venue_state": "QLD",
          "video": {
            "channel": {
              "channel_id": "bd93ca0c-63fc-4bf3-b5b1-efcd208546dc",
            },
          },
        },
        {
          "advertised_start": "2023-03-01T03:38:00Z",
          "category_id": "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
          "meeting_id": "1e690194-59f2-4dfb-b5bc-cf548601bfb4",
          "meeting_name": "Bendigo",
          "race_form": {
            "additional_data": "{"revealed_race_info":{"track_name":"Bendigo","state":"VIC","country":"AUS","number":1,"race_name":"","time":"1970-01-01T00:00:00Z"}}",
            "distance": 425,
            "distance_type": {
              "name": "Metres",
              "short_name": "m",
            },
            "race_comment": "BROADSIDE (2) placed last start at this track when solid behind Colour Me (24.61) in a Maiden. Stalked the pace but found the line strongly. Must come into calculations. MARIONVALE GAL (5) is still a maiden after 17 runs but rarely far away, the latest filling a minor placing when third at Healesville. Has placed at her last two starts now. Expecting another good run. DISTANCE RULE (8) made a promising debut when hitting the frame, finishing third (23.23) at Warragul over 400m. Expected to do much better this time. REDSHIFT VICTORY (1) is right in the mix.",
            "track_condition": {
              "name": "Good",
              "short_name": "good",
            },
            "weather": {
              "name": "Clear",
            },
          },
          "race_id": "059f9e7e-214e-4a34-8480-8014528b15c9",
          "race_name": "Sportsbet Green Ticks (250+ Rank)",
          "race_number": 1,
          "venue_country": "AUS",
          "venue_id": "ab57e842-c2b1-4508-b1c8-dfa908a2db54",
          "venue_name": "Bendigo",
          "venue_state": "VIC",
        },
        {
          "advertised_start": "2023-03-01T03:40:00Z",
          "category_id": "4a2788f8-e825-4d36-9894-efd4baf1cfae",
          "meeting_id": "bac456b9-e41f-47b8-8781-a44d2e435105",
          "meeting_name": "Kawasaki",
          "race_form": {
            "additional_data": "{"revealed_race_info":{"track_name":"Kawasaki","state":"JPN","country":"JPN","number":4,"race_name":"","time":"1970-01-01T00:00:00Z"}}",
            "distance": 2000,
            "distance_type": {
              "name": "Metres",
              "short_name": "m",
            },
            "race_comment": "SILVER SADDLE (7) broke his maiden six starts back and has not been far away since, most recently leading and boxing on for third over 1700m at Funabashi. Takayuki Yano has registered two minor placings when on board. Leading contender. LAMBDA (10) finished fourth last start after being prominent at the bend over 1500m at this track. At peak fitness fourth up and has strong prospects. TSUTSU (5) was in the market but never really a threat when resuming. Won the only start of his previous campaign. Winning claims on best figures. KANAIANZEN (2) ran on to be beaten 3.5 lengths into third at Funabashi two starts ago. Chance on that form.",
            "track_condition": {
              "name": "Good",
              "short_name": "good",
            },
            "weather": {
              "name": "FINE",
            },
          },
          "race_id": "0aad1ddf-b2e5-4680-a1ee-c4cfa85032cf",
          "race_name": "Race 4 - 2000",
          "race_number": 4,
          "venue_country": "JPN",
          "venue_id": "54fd43ae-62e7-4fff-a114-1ebd5378e3a7",
          "venue_name": "Kawasaki",
          "venue_state": "JPN",
        },
        {
          "advertised_start": "2023-03-01T03:42:00Z",
          "category_id": "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
          "meeting_id": "77dbbc03-c138-4774-a2a8-76d6df9d6bd9",
          "meeting_name": "Gawler",
          "race_form": {
            "additional_data": "{"revealed_race_info":{"track_name":"Gawler","state":"SA","country":"AUS","number":9,"race_name":"","time":"1970-01-01T00:00:00Z"}}",
            "distance": 400,
            "distance_type": {
              "name": "Metres",
              "short_name": "m",
            },
            "race_comment": "ARRESTED (7) copped a check in the run and boxed on to place when a 9 length third (23.56) in a Grade 6 at this track. Looks the one to beat on best. ACHING (2) has a strong winning strike-rate over this track and trip with two wins and a placing from four tries. Below best last time at this track (23.25) in a Special conditions race, finishing a 11 length fourth. Can hit the frame with the right run. BETTY'S STAR (4) was a strong winner (23.08) from box 8 last time, scoring by 4 lengths at this track. Has the squeeze box to overcome and will need the breaks early. Significant danger. MAC FRENZY (5) was out of the money here last time out but can bounce back.",
            "track_condition": {
              "name": "Good",
              "short_name": "good",
            },
            "weather": {
              "name": "Cloudy",
            },
          },
          "race_id": "26a1d8a7-773e-48b7-84ea-5f7e56197a41",
          "race_name": "Greyhounds As Pets Stake Pr2 Division5",
          "race_number": 9,
          "venue_country": "AUS",
          "venue_id": "8e36b563-f688-4187-8a98-07633bdc0769",
          "venue_name": "Gawler",
          "venue_state": "SA",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingVertical": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontWeight": "600",
                    },
                    {
                      "marginRight": 4,
                    },
                  ]
                }
              >
                Yonkers Raceway (USA)
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/images/trots.png",
                  }
                }
                style={
                  {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "#0064b2",
                }
              }
            >
              Race 11
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "fontWeight": "600",
                },
                {
                  "color": "#ea5220",
                },
              ]
            }
          >
            0s
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e0e0e0",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingVertical": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontWeight": "600",
                    },
                    {
                      "marginRight": 4,
                    },
                  ]
                }
              >
                Doomben (QLD)
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/images/horses.png",
                  }
                }
                style={
                  {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "#0064b2",
                }
              }
            >
              Race 1
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "fontWeight": "600",
                },
                {
                  "color": "#ea5220",
                },
              ]
            }
          >
            0s
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e0e0e0",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingVertical": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontWeight": "600",
                    },
                    {
                      "marginRight": 4,
                    },
                  ]
                }
              >
                Bendigo (VIC)
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/images/dogs.png",
                  }
                }
                style={
                  {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "#0064b2",
                }
              }
            >
              Race 1
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "fontWeight": "600",
                },
                {
                  "color": "#ea5220",
                },
              ]
            }
          >
            0s
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e0e0e0",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingVertical": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontWeight": "600",
                    },
                    {
                      "marginRight": 4,
                    },
                  ]
                }
              >
                Kawasaki (JPN)
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/images/horses.png",
                  }
                }
                style={
                  {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "#0064b2",
                }
              }
            >
              Race 4
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "fontWeight": "600",
                },
                {
                  "color": "#ff7800",
                },
              ]
            }
          >
            2m 0s
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#e0e0e0",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingVertical": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontWeight": "600",
                    },
                    {
                      "marginRight": 4,
                    },
                  ]
                }
              >
                Gawler (SA)
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/images/dogs.png",
                  }
                }
                style={
                  {
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
            <Text
              style={
                {
                  "color": "#0064b2",
                }
              }
            >
              Race 9
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "fontWeight": "600",
                },
                {
                  "color": "#000000",
                },
              ]
            }
          >
            4m
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
